### ISSUE TITLE
💯  Minor codebase updates and fixes

### ISSUE THREAD

**Issue by fregante** [refined-github/refined-github#4008]
I'll use this issue to add small tasks that could be done, treat it as a wishlist or a good way to start getting acquainted with the codebase.

- [ ] Find fickle selectors and change them with `select(something more specific).closest(the parent)`;

	GitHub has been removing some feature-specific classes like `.js-file-list` which means that we end up with monstrosities like `main > :first-child .mt-1 .btn` which mean nothing and could change at any time. Ideally instead we should find the best selector inside the target element (for example its icon, `select('.octicon-branch[aria-label="Create branch"]')`) and then reach the desired element with `.closest('.btn')`:

  ```diff
  - select('main > :first-child .mt-1 .btn')
  + select('.octicon-branch[aria-label="Create branch"]')!.closest('.btn')
  ```

- Better yet: https://github.com/sindresorhus/refined-github/issues/4456

**Comment by doc-han** [comment]
Hi, I'm taking on this

**Comment by fregante** [comment]
Thank you! They’re all different points, so please do one PR for each so it’s easier to review.

**Comment by cheap-glitch** [comment]
> Use mem.decorator wherever possible sindresorhus/mem#61

Sadly, TypeScript decorators can only be used inside classes: https://www.typescriptlang.org/docs/handbook/decorators.html#decorators

**Comment by fregante** [comment]
- [ ] Change every GHE-related comments to `/* GHE #issue */` so that they're easily trackable and removed once old enough. Never "GitHub Enterprise" but always GHE (example: https://github.com/sindresorhus/refined-github/blob/3aadf2f9141107d7ca92e8753f2a66cbc10ebd9d/source/features/safer-destructive-actions.css#L4)
- [ ] Possibly review and remove such comments if they're older than 1 year

**Comment by yakov116** [comment]
The last 2 I have included in my june refresh PR (un committed). If you want I can commit them now. But I would rather wait for then so we know that we are breaking GHE in one PR 😄

**Comment by fregante** [comment]
- [ ] use `.base-ref` instead of `:not(.head-ref)` 🤷‍♂️ https://github.com/sindresorhus/refined-github/blob/1a0d254f87e235dd7991c1b8f4034326dc822800/source/features/clean-conversation-headers.tsx#L32:L32

**Comment by yakov116** [comment]
Change 
https://github.com/sindresorhus/refined-github/blob/ee8b6422d2ddd1199f557594d03e7ee850e843f0/source/features/enable-file-links-in-compare-view.tsx#L62-L63

To `pageDetect.isClosedPR`

**Comment by fregante** [comment]
- [ ] Move negative exclusions to their own list of `required` tests

https://github.com/sindresorhus/refined-github/blob/936eedbb0bb37a30fc44a2cc7e5f3af31f24a3fc/source/features/fork-source-link-same-view.tsx#L24-L32

Could be

```js

	include: [
		pageDetect.isSingleFile,
		pageDetect.isRepoTree,
		pageDetect.isEditingFile
	],
	required: [ // name TBD
		pageDetect.isForkedRepo
	],
	exclude: [
		pageDetect.isRepoRoot
	],
```

**Comment by fregante** [comment]
- [ ] Maybe rewrite most `selector-observer` usage to use `deinit` instead of `onetime` as described in https://github.com/sindresorhus/refined-github/pull/4072

**Comment by fregante** [comment]
- [ ] Use https://github.com/g-plane/typed-query-selector in fetch-dom

**Comment by fregante** [comment]
- [ ] These should be 4 checkboxes as originally intended:

<img width="593" alt="Screen Shot 2021-04-04 at 00 46 42" src="https://user-images.githubusercontent.com/1402241/113486909-40295300-94df-11eb-8efd-98b1a570d00c.png">

This makes it clearer to users that a no-scope token is still fine for read-only operations. Currently instead no error is shown, but the user still sees three ❌

**Comment by yakov116** [comment]
Replace 
https://github.com/sindresorhus/refined-github/blob/a94b69550dde2e9a75001470dc75d17e6cf50401/source/features/table-input.tsx#L51-L53

With `octoicon-table` https://github.com/primer/octicons/releases/tag/v13.0.0

**Comment by cheap-glitch** [comment]
Seems like GitHub moved the tooltips orientation in the editor's toolbar, and now RGH's tooltips don't match:
![image](https://user-images.githubusercontent.com/46634000/114894352-daa66200-9e0e-11eb-9866-79e08bd90013.png)
Worth fixing?

**Comment by fregante** [comment]
- [ ] Replace `exclude: [isEmptyRepo]` and related exclusions with a check in the error catcher in features/index. Instead of having to explicitly specify these exclusions, RG should just understand that the current page is incomplete and ignore every error. Maybe it should just show once a log: “This page is incomplete. Some features have been disabled.”

The feature logger might also display a log with a 💢 emoji or similar.

**Comment by fregante** [comment]
- Use `typed-query-selector` for `closest` as well, follows https://github.com/sindresorhus/refined-github/pull/3831

This means we can stop using unnecessary generics like `.closest<HMTLElement>('.hi')`

**Comment by sindresorhus** [comment]
- [ ] Add a direct "rate" link to the options page, to encourage the user to rate the extension in the Mac App Store / Chrome Store / Firefox Addons.

**Comment by cheap-glitch** [comment]
Maybe this:
https://github.com/sindresorhus/refined-github/blob/2d165c692e2fcdd5b0a01a01ee0bc37af7acefd1/source/features/clean-repo-sidebar.css#L13
could be replaced by `.mt-n3` (source: https://primer.style/css/utilities/margin#negative-margins)

**Comment by fregante** [comment]
Exclusively if we already create/alter the element via JS

**Comment by fregante** [comment]
- [ ] Add `updateRepoNavigation` function export to `more-dropdown` so that other features like `releases-tab` and `bugs-tab` can call it when they update the content. Occasionally I still see https://github.com/sindresorhus/refined-github/issues/3347, probably due to `bugs-tab`

**Comment by yakov116** [comment]
@fregante what updates should I know about for typescript 4.3?

**Comment by fregante** [comment]
- [ ] Add workflow to ensure that PRs with titles like `add xyz feature`, include a file called `xyz.tsx` and add a line with `xyz` in `readme.md`. If they aren't found in the diff, it should fail. This fixes https://github.com/sindresorhus/refined-github/issues/4320#issuecomment-833301366

> @fregante what updates should I know about for typescript 4.3?

🤷‍♂️

**Comment by cheap-glitch** [comment]
- [ ] Use `on-details-menu-selected` in conversation-activity-filter and remove the `await delay(1)` in the event handler (_originally mentioned in https://github.com/sindresorhus/refined-github/pull/4526#discussion_r660522833_)

**Comment by yakov116** [comment]
Add https://github.blog/changelog/2021-07-02-github-actions-setup-node-now-supports-dependency-caching?

**Comment by fregante** [comment]
- [ ] Replace manual deduplication checks with `deduplicate`

https://github.com/sindresorhus/refined-github/blob/40ed43f6cdd23178b6c6ee7c04a5731cb2342012/source/features/clean-conversation-sidebar.tsx#L60-L63

This should just be `deduplicate: '.rgh-clean-sidebar'` I think

**Comment by yakov116** [comment]
Nope deduplucate will not work on `additionalListeners`

**Comment by fregante** [comment]
Darn it, again 😃

**Comment by fregante** [comment]
- [ ] Add a general spellchecker? I wish there was a simple solution to spellcheck text in comments, markdown and ideally even in phrase-like strings. Example: https://github.com/marketplace/actions/github-spellcheck

I think that tool lets us extend the dictionary by editing a file at the root of the repo. It would be ok if it worked out of the box or with a dozen of new words in the dictionary, but if it requires us to add a hundred words then it’s not worth it anymore.

**Comment by fregante** [comment]
- [ ] Fix some TODOs 😅

**Comment by fregante** [comment]
_From https://github.com/sindresorhus/refined-github/issues/4197#issuecomment-820954304_

- [ ] Rewrite `copy-file` so that it uses the native `clipboard-copy` element.

I'm not sure if this is possible, but I'm imagining that if we use a "capture: true" we can run before their event and set the required attribute:

1. Onclick -> `button.setAttribute('value', blob.innerText)`
2. `clipboard-copy` takes care of showing the tooltip and actually copying to the clipboard

Also our onclick handler only needs to run once (if this optimization is possible)

**Comment by fregante** [comment]
- [ ] Fix every instance of `unicorn/no-array-callback-reference` and make it an error in XO’s config

**Comment by kidonng** [comment]
Can I post here as well? Would like to add back the username in `reaction-avatars`: https://github.com/sindresorhus/refined-github/pull/4608#discussion_r678439504

**Comment by fregante** [comment]
You can, but this issue is about refactoring and other meta improvements, not fixes/changes to specific features. That specific tooltip was removed at some point, so it's worth blaming the line and opening a new issue, if you want.

**Comment by fregante** [comment]
- [ ] Add `.prettierignore` file to disable "prettier on save" feature of every editor: https://prettier.io/docs/en/ignore.html

The content would probably be just `*.*` but it needs to be tested

**Comment by fregante** [comment]
- [ ] Probably revert https://github.com/gamemaker1/refined-github/commit/bfdad4b4ea240c93ab0c51ff4e2a71a5b3849c94

The solution seems awkward, it needs to be reconsidered. Blocked by:

- #

**Comment by cheap-glitch** [comment]
Noticed a small visual bug in `list-prs-for-file` (Firefox):
![image](https://user-images.githubusercontent.com/46634000/136665914-cea0e4bc-dbae-440f-8870-a2ffde09893b.png)

**Comment by fregante** [comment]
- [ ]  replace `__basefilename` with a tiny function like `getFilename(import.meta.url)`

**Comment by fregante** [comment]
- [ ] Switch to esbuild

Now that these are in, there's little keeping us on Webpack

- https://github.com/refined-github/refined-github/pull/5058
- #5049

https://esbuild.github.io/api/#advanced-options

**Comment by fregante** [comment]
- [ ] Fix `stripIndent` during issue link creation

This is a link it created. You can see that `stripIndent` failed because the stack actually wraps and a new line starts without indentation, making stripIndent think there's no indentation to be dropped.

Its documentation should mention a fix

https://github.com/refined-github/refined-github/issues/new?labels=bug&template=1_bug_report.yml&title=%60conversation-filters%60%3A+sourceItem+is+undefined&example_url=https%3A%2F%2Fgithub.com%2Frefined-github%2Frefined-github%2Fissues%3Fq%3Dis%253Aissue%2Bis%253Aopen%2Bsort%253Aupdated-desc&description=%0A%09%09%60%60%60%0A%09%09init%40moz-extension%3A%2F%2F975e98b4-cdf5-cb4f-bf14-7a9ca1ba5e21%2Fbuild%2Frefined-github.js%3A5091%3A24%0ArunFeature%40moz-extension%3A%2F%2F975e98b4-cdf5-cb4f-bf14-7a9ca1ba5e21%2Fbuild%2Frefined-github.js%3A7988%3A17%0AsetupPageLoad%40moz-extension%3A%2F%2F975e98b4-cdf5-cb4f-bf14-7a9ca1ba5e21%2Fbuild%2Frefined-github.js%3A8006%3A11%0A.%2Fsource%2Ffeatures%2Findex.tsx%2Fadd%2F%3C%40moz-extension%3A%2F%2F975e98b4-cdf5-cb4f-bf14-7a9ca1ba5e21%2Fbuild%2Frefined-github.js%3A8061%3A15%0Aasync*add%40moz-extension%3A%2F%2F975e98b4-cdf5-cb4f-bf14-7a9ca1ba5e21%2Fbuild%2Frefined-github.js%3A8062%3A9%0Aasync*.%2Fsource%2Ffeatures%2Fconversation-filters.tsx%40moz-extension%3A%2F%2F975e98b4-cdf5-cb4f-bf14-7a9ca1ba5e21%2Fbuild%2Frefined-github.js%3A5106%3A48%0A__webpack_require__%40moz-extension%3A%2F%2F975e98b4-cdf5-cb4f-bf14-7a9ca1ba5e21%2Fbuild%2Frefined-github.js%3A25560%3A42%0A%40moz-extension%3A%2F%2F975e98b4-cdf5-cb4f-bf14-7a9ca1ba5e21%2Fbuild%2Frefined-github.js%3A25671%3A108%0A%40moz-extension%3A%2F%2F975e98b4-cdf5-cb4f-bf14-7a9ca1ba5e21%2Fbuild%2Frefined-github.js%3A26026%3A3%0A%40moz-extension%3A%2F%2F975e98b4-cdf5-cb4f-bf14-7a9ca1ba5e21%2Fbuild%2Frefined-github.js%3A26028%3A12%0A%0A%09%09%60%60%60%0A%09

**Comment by fregante** [comment]
- [ ] Fix: Safari error messages don't include the header

For some reason only in Firefox we only get the X emoji but nothing after it.

### Current

<img width="562" alt="Screen Shot 7" src="https://user-images.githubusercontent.com/1402241/145542477-1587c0bd-9782-4594-a912-c96a5ce686a4.png">


### Expected

<img width="775" alt="Screen Shot 9" src="https://user-images.githubusercontent.com/1402241/145542653-309a4e5d-b936-4589-b5b3-75347a18e129.png">

**Comment by fregante** [comment]
Let's not add any more requests to this issue. A new one can be created. We'll close this one after (almost) everything is done. 

# 🛁

**Comment by FloEdelmann** [comment]
> Replace `fregante/release-with-changelog` with other action that uses GitHub’s autogenerator (if it exists).

The ["create release" REST API endpoint](https://docs.github.com/en/rest/reference/releases#create-a-release) accepts a `generate_release_notes` option which can be called in a workflow: https://github.com/MylesBorins/node-osc/blob/v6.1.11/.github/workflows/create-release.yml#L26-L41

**Comment by fregante** [comment]
Thanks! I opened https://github.com/fregante/release-with-changelog/issues/53

As far as RG is concerned, we can copy-paste the lines you suggested for now as the issue I just opened is more involved.

**Comment by FloEdelmann** [comment]
Regarding https://github.com/refined-github/refined-github/issues/4008#issuecomment-824729016:

The specific linked case doesn't exist anymore. Out of all other currently existing negative margins/paddings, only one of two in `clean-mergeability-box` could be replaced with a class, but that is a CSS-only feature. All others are not compatible with Primer classes.

**Comment by fregante** [comment]
Closing and locking. Thanks for all the fish!
